[WIP] Add AdaptiveAvgPool to test lightweight dispatch

priyaramani opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73948

Differential Revision: [D34740736](https://our.internmc.facebook.com/intern/diff/D34740736)

priyaramani committed:
Add AdaptiveAvgPool to test lightweight dispatch

[ghstack-poisoned]

priyaramani committed:
Update on "Add AdaptiveAvgPool to test lightweight dispatch"




[ghstack-poisoned]

priyaramani committed:
Update on "Add AdaptiveAvgPool to test lightweight dispatch"




[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34740736).

larryliu0820 commented on aten/src/ATen/core/Tensor.cpp:
```diff
@@ -28,7 +29,7 @@ TensorBase TensorBase::__dispatch_contiguous(c10::MemoryFormat memory_format) co
 
```
It would be weird to not cover `contiguous` and `zero_` if we are going to change `fill_`. I suggest let's add tests to cover for these two ops as well.

ezyang commented on aten/src/ATen/core/Tensor.cpp:
```diff
@@ -28,7 +29,7 @@ TensorBase TensorBase::__dispatch_contiguous(c10::MemoryFormat memory_format) co
 const TensorBase& TensorBase::fill_(const c10::Scalar &fill_value) const {
   Tensor self(*this);
-  at::_ops::fill__Scalar::call(self, fill_value);
+  at::fill_(self, fill_value);
```
This doesn't look right. cc @peterbell10

larryliu0820 replied:
Hi Edward, for some context, this change is to allow lightweight dispatch build to use `fill_` `contiguous` and `zero_` in `Tensor.cpp`. The current call to `at::_ops::fill__Scalar::call` will visit dispatcher which will not be available in lightweight dispatch. On the other hand, we can call `at::fill_` which codegen provides two different implementations.

With static dispatch enabled:
```
// aten::fill_.Tensor(Tensor(a!) self, Tensor value) -> Tensor(a!)
TORCH_API inline at::Tensor & fill_(at::Tensor & self, const at::Tensor & value) {
    return at::cpu::fill_(self, value);
}
```

With static dispatch disabled:
```
// aten::fill_.Tensor(Tensor(a!) self, Tensor value) -> Tensor(a!)
TORCH_API inline at::Tensor & fill_(at::Tensor & self, const at::Tensor & value) {
    return at::_ops::fill__Tensor::call(self, value);
}
```

Let us know if you have better idea to support both options.

peterbell10 replied:
Would it be possible for `at::_ops::foo::call` to use static dispatch? That way only one call site needs to be updated and `Tensor` and `at::foo` benefit automatically, as well as any other users of the ops interface.

larryliu0820 replied:
@peterbell10 it seems like a big change. So from my shallow understanding, `foo_ops.h` is a per-header version of `Operators.cpp`. I think in some use cases, we still need to go through the dispatcher as well as static dispatch, thus we need both `foo_ops.h` and `foo.h`.

peterbell10 replied:
If there are situations where the both static and dynamic dispatch are used together then I agree that wouldn't work. I didn't realize that could happen.

cccclai replied:
Just see the pr - what will be the case to still go through the dispatcher when specify static dispatch? How do we know which operators will go through dispatcher or static dispatch?

priyaramani committed:
Update on "[WIP] Add AdaptiveAvgPool to test lightweight dispatch"


Differential Revision: [D34740736](https://our.internmc.facebook.com/intern/diff/D34740736)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34740736).

priyaramani committed:
Update on "[WIP] Add AdaptiveAvgPool to test lightweight dispatch"


Differential Revision: [D34740736](https://our.internmc.facebook.com/intern/diff/D34740736)

[ghstack-poisoned]

priyaramani committed:
Update on "[WIP] Add AdaptiveAvgPool to test lightweight dispatch"


Differential Revision: [D34740736](https://our.internmc.facebook.com/intern/diff/D34740736)

[ghstack-poisoned]

priyaramani committed:
Update on "[WIP] Add AdaptiveAvgPool to test lightweight dispatch"


Differential Revision: [D34740736](https://our.internmc.facebook.com/intern/diff/D34740736)

[ghstack-poisoned]
